include/hobbes/eval/cc.H

```diff
@@ -84,175 +99,205 @@ public:
+  typedef ExprPtr (*readExprFn)(cc *, const std::string &);
+  ExprPtr readExpr(const std::string &);
+  MonoTypePtr readMonoType(const std::string &);
   void setReadExprFn(readExprFn);
 
   struct UnreachableMatches {
     LexicalAnnotation la;
     std::string lines;
   };
-  typedef void (*gatherUnreachableMatchesFn)(const UnreachableMatches&);
-  void gatherUnreachableMatches(const UnreachableMatches& m);
+  typedef void (*gatherUnreachableMatchesFn)(const UnreachableMatches &);
+  void gatherUnreachableMatches(const UnreachableMatches &m);
   void setGatherUnreachableMatchesFn(gatherUnreachableMatchesFn f);
 
   // search for paths from one type to another
   // (currently just one-step paths, may be useful to consider multi-step paths)
-  SearchEntries search(const MonoTypePtr&, const MonoTypePtr&);
-  SearchEntries search(const ExprPtr&, const MonoTypePtr&);
-  SearchEntries search(const std::string&, const MonoTypePtr&);
-  SearchEntries search(const std::string&, const std::string&);
+  SearchEntries search(const MonoTypePtr &, const MonoTypePtr &);
+  SearchEntries search(const ExprPtr &, const MonoTypePtr &);
+  SearchEntries search(const std::string &, const MonoTypePtr &);
+  SearchEntries search(const std::string &, const std::string &);
+
 private:
   // parser functions
   readModuleFileFn readModuleFileF;
-  readModuleFn     readModuleF;
-  readExprDefnFn   readExprDefnF;
-  readExprFn       readExprF;
+  readModuleFn readModuleF;
+  readExprDefnFn readExprDefnF;
+  readExprFn readExprF;
 
-  gatherUnreachableMatchesFn gatherUnreachableMatchesF = [](const UnreachableMatches&){};
+  gatherUnreachableMatchesFn gatherUnreachableMatchesF =
+      [](const UnreachableMatches &) {};
 
 public:
   // type-safe compilation to C++ function pointers
-  template <typename RFn, typename ... NamesAndExpr>
-    typename func<RFn>::type compileFn(NamesAndExpr ... args) {
-      static_assert(func<RFn>::arity == sizeof...(NamesAndExpr)-1, "Formal parameter list and expected function type arity mismatch");
-      return rcast<typename func<RFn>::type>(
-        unsafeCompileFn(
-          lift<RFn>::type(*this),
-          PArgl<NamesAndExpr...>::names(args...),
-          PArgl<NamesAndExpr...>::expr([&](const std::string& x){return this->readExpr(x);}, args...)
-        )
-      );
-    }
+  template <typename RFn, typename... NamesAndExpr>
+  typename func<RFn>::type compileFn(NamesAndExpr... args) {
+    static_assert(
+        func<RFn>::arity == sizeof...(NamesAndExpr) - 1,
+        "Formal parameter list and expected function type arity mismatch");
+    return rcast<typename func<RFn>::type>(unsafeCompileFn(
+        lift<RFn>::type(*this), PArgl<NamesAndExpr...>::names(args...),
+        PArgl<NamesAndExpr...>::expr(
+            [&](const std::string &x) { return this->readExpr(x); }, args...)));
+  }
+
+  template <typename RFn, typename... NamesAndExpr>
+  boolSafeFn<RFn> compileSafeFn(NamesAndExpr &&...args) {
```

Comment:
@mo-xiaoming , @dawa79 - please help review this fix

Reply:
let's assess later (as a separate issue) what it would take to conflate `compileFn` and `compileSafeFn` into one single function